R/utilities.R

```diff
@@ -509,6 +509,191 @@ ExpMean <- function(x) {
 }
 
+#' Export Seurat object for UCSC cell browser
+#'
+#' @param object Seurat object
+#' @param dir output directory path
+#' @param dataset.name name of the dataset
+#' @param vars vector of metadata fields to export
+#' @param vars.names list that defines metadata field names after the export.
+#' Should map metadata column name to export name
+#' @param reductions vector of reduction names to export
+#' @param markers.file path to file with marker genes
+#' @param cluster.field name of the metadata field containing cell cluster
+#' @param cb.dir in which dir to create UCSC cell browser content root
+#' @param port \emph{experimental} on which port to run UCSC cell browser after export
+#'
+#' @return ...
+#'
+#' @author Maximilian Haeussler, Nikolay Markov
+#'
+#' @importFrom utils browseURL
+#' @importFrom reticulate py_module_available import
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' ExportToCellbrowser(object = pbmc_small, dataset.name = "PBMC", dir = "out")
+#' }
+#'
+ExportToCellbrowser <- function(
+  object,
+  dir,
+  dataset.name,
+  vars = NULL,
+  vars.names = NULL,
+  reductions = "tsne",
+  markers.file = NULL,
+  cluster.field = NULL,
+  port = NULL,
+  cb.dir = NULL
+) {
+  if (is.null(x = vars)) {
+    vars <- c("nCount_RNA", "nFeature_RNA")
+    if (length(x = levels(x = Idents(object = object))) > 1) {
+      vars <- c(vars, ".ident")
+    }
+  }
+  if (!is.null(x = port) && is.null(x = cb.dir)) {
+    stop("cb.dir parameter is needed when port is set")
+  }
+  if (!dir.exists(paths = dir)) {
+    dir.create(path = dir)
+  }
+  if (!dir.exists(paths = dir)) {
+    stop("Output directory ", dir, " cannot be created or is a file")
+  }
+  order <- colnames(x = object)
+  enum.fields <- c()
+  # Export expression matrix
+  df <- as.data.frame(x = as.matrix(x = GetAssayData(object = object)))
+  df <- data.frame(gene = rownames(x = object), df)
+  z <- gzfile(file.path(dir, "exprMatrix.tsv.gz"), "w")
+  write.table(x = df, sep = "\t", file = z, quote = FALSE, row.names = FALSE)
+  close(con = z)
+  # Export cell embeddings
+  embeddings.conf <- c()
+  for (reduction in reductions) {
+    df <- Embeddings(object = object, reduction = reduction)
+    if (ncol(x = df) > 2) {
+      warning(
+        'Embedding ',
+        reduction,
+        ' has more than 2 coordinates, taking only the first 2'
+      )
+      df <- df[, 1:2]
+    }
+    colnames(x = df) <- c("x", "y")
+    df <- data.frame(cellId = rownames(x = df), df)
+    fname <- file.path(dir, paste0(reduction, '.coords.tsv'))
+    write.table(
+      x = df[order, ],
+      sep = "\t",
+      file = fname,
+      quote = FALSE,
+      row.names = FALSE
+    )
+    conf <- sprintf(
+      '{"file": "%s.coords.tsv", "shortLabel": "Seurat %1$s"}',
+      reduction
+    )
+    embeddings.conf <- c(embeddings.conf, conf)
+  }
+  # Export metadata
+  df <- data.frame(row.names = rownames(x = object[[]]))
+  for (field in vars) {
```

Comment:
Could this for-loop be replaced by a call to `FetchData`, especially as passing `'ident'` will cause `FetchData` to pull the active identity classes.